Mathlib/NumberTheory/Zagier.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2023 Jeremy Tan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Tan, Thomas Browning
+-/
+import Mathlib.GroupTheory.Perm.Cycle.Type
+
+/-!
+# Zagier's "one-sentence proof" of Fermat's theorem on sums of two squares
+
+"The involution on the finite set `S = {(x, y, z) : ℕ × ℕ × ℕ | x ^ 2 + 4 * y * z = p}` defined by
+```
+(x, y, z) ↦ (x + 2 * z, z, y - x - z) if x < y - z
+            (2 * y - x, y, x - y + z) if y - z < x < 2 * y
+            (x - 2 * y, x - y + z, y) if x > 2 * y
+```
+has exactly one fixed point, so `|S|` is odd and the involution defined by
+`(x, y, z) ↦ (x, z, y)` also has a fixed point."
+-/
+
+
+namespace Zagier
+
+section Sets
+
+open Finset
+
+variable (k : ℕ) [hk : Fact (4 * k + 1).Prime]
+
+/-- The set of all triples of natural numbers `(x, y, z)` satisfying
+`x * x + 4 * y * z = 4 * k + 1`, as a `Set`. -/
+def zagierSet : Set (ℕ × ℕ × ℕ) := {t | t.1 * t.1 + 4 * t.2.1 * t.2.2 = 4 * k + 1}
+
+lemma zagierSet_lower_bound {x y z : ℕ} (h : (x, y, z) ∈ zagierSet k) : 0 < x ∧ 0 < y ∧ 0 < z := by
+  simp_rw [zagierSet, Set.mem_setOf_eq] at h
+  refine' ⟨_, _, _⟩ <;> (by_contra q; push_neg at q; rw [le_zero_iff] at q; subst q; simp at h)
+  · apply_fun (· % 4) at h
+    simp [mul_assoc, Nat.add_mod] at h
+  all_goals cases' (Nat.dvd_prime hk.out).1 (dvd_of_mul_left_eq _ h) with e e <;>
+    (subst e; simp at h; subst h; simp at hk; exact hk.out)
+
+lemma zagierSet_upper_bound {x y z : ℕ} (h : (x, y, z) ∈ zagierSet k) :
+    x ≤ k + 1 ∧ y ≤ k ∧ z ≤ k := by
+  obtain ⟨hx, hy, hz⟩ := zagierSet_lower_bound k h
+  simp_rw [zagierSet, Set.mem_setOf_eq] at h
+  refine' ⟨_, _, _⟩ <;> nlinarith
+
+/-- The set of all triples of natural numbers `(x, y, z) ∈ (0, k + 1] × (0, k] × (0, k]` satisfying
+`x * x + 4 * y * z = 4 * k + 1`, as a `Finset`. This is shown to be equivalent to `zagierSet`. -/
+def zagierFinset : Finset (ℕ × ℕ × ℕ) :=
+  ((Ioc 0 (k + 1)).product ((Ioc 0 k).product (Ioc 0 k))).filter
+    (fun ⟨x, y, z⟩ => x * x + 4 * y * z = 4 * k + 1)
+
+theorem coe_zagierFinset : zagierFinset k = zagierSet k := by
+  ext ⟨x, y, z⟩
+  refine' ⟨fun h => _, fun h => _⟩
+  · unfold zagierFinset at h; unfold zagierSet; simp_all
+  · unfold zagierSet at h; unfold zagierFinset; simp_all
+    have lb := zagierSet_lower_bound k h
+    have ub := zagierSet_upper_bound k h
+    apply mem_product.2; simp
+    constructor; · exact ⟨lb.1, ub.1⟩
+    apply mem_product.2; simp
+    exact ⟨⟨lb.2.1, ub.2.1⟩, ⟨lb.2.2, ub.2.2⟩⟩
+
+instance : Fintype (zagierSet k) := by rw [← coe_zagierFinset]; infer_instance
```

Comment:
```suggestion
lemma zagierSet_subset : zagierSet k ⊆ Ioc 0 (k + 1) ×ˢ Ioc 0 k ×ˢ Ioc 0 k := by
  intro x h
  have lb := zagierSet_lower_bound k h
  have ub := zagierSet_upper_bound k h
  exact ⟨⟨lb.1, ub.1⟩, ⟨lb.2.1, ub.2.1⟩, ⟨lb.2.2, ub.2.2⟩⟩

noncomputable instance : Fintype (zagierSet k) :=
  (((finite_Ioc 0 (k + 1)).prod
    ((finite_Ioc 0 k).prod (finite_Ioc 0 k))).subset (zagierSet_subset k)).fintype
```

Reply:
Done, with a change from `open Finset` to `open Set` (otherwise `Ioc` causes errors due to `Set.Ioc` being required and not `Finset.Ioc`).

Reply:
Oh, yes, I forgot about that change.